compiler/src/jdk.graal.compiler/src/jdk/graal/compiler/phases/common/WriteBarrierAdditionPhase.java

```diff
@@ -37,19 +37,29 @@
 
 public class WriteBarrierAdditionPhase extends BasePhase<CoreProviders> {
 
+    private final StageFlag stage;
+
+    public WriteBarrierAdditionPhase() {
+        this(StageFlag.BARRIER_ADDITION);
+    }
+
+    public WriteBarrierAdditionPhase(StageFlag stage) {
+        this.stage = stage;
+    }
+
     @Override
     public Optional<NotApplicable> notApplicableTo(GraphState graphState) {
         return NotApplicable.ifAny(
-                        NotApplicable.ifApplied(this, StageFlag.BARRIER_ADDITION, graphState),
-                        NotApplicable.unlessRunAfter(this, StageFlag.MID_TIER_LOWERING, graphState),
+                        NotApplicable.ifApplied(this, stage, graphState),
+                        NotApplicable.unlessRunAfter(this, stage == StageFlag.BARRIER_ADDITION ? StageFlag.MID_TIER_LOWERING : StageFlag.LOW_TIER_LOWERING, graphState),
```

Comment:
I think this would look nicer if we renamed `BARRIER_ADDITION` to `MID_TIER_BARRIER_ADDITION` for symmetry. The constructor should check that the stage flag passed in is one of `MID_TIER_BARRIER_ADDITION` or `LOW_TIER_BARRIER_ADDITION`.